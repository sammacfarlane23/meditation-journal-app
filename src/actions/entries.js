import database from '../firebase/firebase';

export const createEntry = (entry) => ({
  type: 'CREATE_ENTRY',
  entry,
});

export const startCreateEntry = (entryData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const { entryText = '', createdAt = 0, title = '' } = entryData;
    const entry = {
      entryText,
      createdAt,
      title,
    };

    return database
      .ref(`users/${uid}/entries`)
      .push(entry)
      .then((ref) => {
        dispatch(
          createEntry({
            id: ref.key,
            ...entry,
          })
        );
      });
  };
};

export const removeEntry = ({ id } = {}) => ({
  type: 'REMOVE_ENTRY',
  id,
});

export const startRemoveEntry = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/entries/${id}`)
      .remove()
      .then(() => {
        dispatch(removeEntry({ id }));
      });
  };
};

export const editEntry = (id, updates) => ({
  type: 'EDIT_ENTRY',
  id,
  updates,
});

export const startEditEntry = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/entries/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editEntry(id, updates));
      });
  };
};

export const setEntries = (entries) => {
  return {
    type: 'SET_ENTRIES',
    entries,
  };
};

export const startSetEntries = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/entries`)
      .once('value')
      .then((snapshot) => {
        const entries = [];

        snapshot.forEach((childSnapshot) => {
          entries.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setEntries(entries));
      });
  };
};

export const selectEntryForEdit = (entry) => {
  return {
    type: 'SELECT_ENTRY_FOR_EDIT',
    entry,
  };
};
