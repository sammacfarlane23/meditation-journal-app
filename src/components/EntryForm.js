import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import EditButton from './EditButton';

export default class EntryForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      entryText: props.entry ? props.entry.entryText : '',
      createdAt: props.entry ? moment(props.entry.createdAt) : moment(),
      error: '',
      calendarFocused: false,
      title: props.entry ? props.entry.title : '',
      editDate: false,
    };
  }

  onEntryTextChange = (e) => {
    const entryText = e.target.value;
    this.setState(() => ({ entryText }));
  };

  onRemove = () => {
    this.props.onRemove(this.props.entry);
  };

  onTitleChange = (e) => {
    const title = e.target.value;
    this.setState(() => ({ title }));
  };

  onTimeChange = (e) => {
    const currentDate = moment(this.state.createdAt).startOf('day');
    const timeOfDay = moment.duration(
      moment(e.target.value, 'HH:mm') -
        moment(e.target.value, 'HH:mm').startOf('day')
    );
    const createdAt = moment(currentDate).add(timeOfDay);
    this.setState(() => ({ createdAt }));
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onDateChange = (createdAt) => {
    const timeOfDay =
      moment(this.state.createdAt) -
      moment(this.state.createdAt).startOf('day');
    const addTime = moment(moment(createdAt).startOf('day') + timeOfDay);
    if (createdAt) {
      this.setState(() => ({ createdAt: addTime }));
    }
  };

  onEditDate = () => {
    this.setState(() => ({ editDate: true }));
  };

  closeModal = () => {
    this.props.closeModal();
  };

  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.entryText) {
      this.setState(() => ({
        error: 'Please add an entry',
      }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        entryText: this.state.entryText,
        createdAt: this.state.createdAt.valueOf(),
        title: this.state.title,
      });
    }
  };

  render() {
    return (
      <form className='form' onSubmit={this.onSubmit}>
        <input
          className='title'
          type='text'
          placeholder='Title (optional)'
          value={this.state.title ? this.state.title : ''}
          onChange={this.onTitleChange}
        />
        <hr className='horizontal-line' />
        {this.state.error && <p className='form__error'>{this.state.error}</p>}
        <textarea
          rows='10'
          value={this.state.entryText}
          className='text-area'
          placeholder='Describe your experience here...'
          onChange={this.onEntryTextChange}
        ></textarea>
        {!this.state.editDate && (
          <div className='createdAt-section'>
            <p className='createdAt-section__text'>
              {moment(this.state.createdAt).format('DD MMM YYYY HH:mm')}
            </p>
            <button
              type='button'
              className='edit-button'
              onClick={this.onEditDate}
            >
              <EditButton />
            </button>
          </div>
        )}
        {this.state.editDate && (
          <div className='date-edit'>
            <div className='single-date-picker'>
              <SingleDatePicker
                date={this.state.createdAt}
                onDateChange={this.onDateChange}
                onFocusChange={this.onFocusChange}
                focused={this.state.calendarFocused}
                numberOfMonths={1}
                isOutsideRange={() => false}
                displayFormat='DD/MM/YYYY'
                isOutsideRange={(d) => d.isAfter(moment())}
                withPortal={true}
              />
            </div>
            <input
              className='input-time'
              type='time'
              onChange={this.onTimeChange}
              value={moment(this.state.createdAt).format('HH:mm')}
              max={
                moment()
                  .startOf('day')
                  .isSame(moment(this.state.createdAt).startOf('day'))
                  ? moment().format('HH:mm')
                  : '23:59'
              }
            />
          </div>
        )}
        <div className='button-container'>
          <div className='remove-button-container'>
            {this.props.canEdit && (
              <button
                type='button'
                className='button button--delete'
                onClick={this.onRemove}
              >
                Delete
              </button>
            )}
          </div>
          <div className='other-button-container'>
            <button className='button button--right'>Save</button>
            <button
              type='button'
              onClick={this.closeModal}
              className='button button--right button--cancel'
            >
              Cancel
            </button>
          </div>
        </div>
      </form>
    );
  }
}
