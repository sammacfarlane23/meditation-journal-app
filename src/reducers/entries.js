const entriesReducerDefaultState = [];

export default (state = entriesReducerDefaultState, action) => {
  switch (action.type) {
    case 'CREATE_ENTRY':
      return [...state, action.entry];
    case 'REMOVE_ENTRY':
      return state.filter(({ id }) => action.id !== id);
    case 'EDIT_ENTRY':
      return state.map((entry) => {
        if (entry.id === action.id) {
          return {
            ...entry,
            ...action.updates,
          };
        } else {
          return entry;
        }
      });
    case 'SET_ENTRIES':
      return action.entries;
    case 'SELECT_ENTRY_FOR_EDIT':
      return action.entry;
    default:
      return state;
  }
};
