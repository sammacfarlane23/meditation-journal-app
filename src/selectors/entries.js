export default (entries, { text }) => {
  return entries
    .filter((entry) => {
      const titleMatch = entry.title.toLowerCase().includes(text.toLowerCase());

      const textMatch = entry.entryText
        .toLowerCase()
        .includes(text.toLowerCase());

      return textMatch || titleMatch;
    })
    .sort((a, b) => {
      return a.createdAt < b.createdAt ? 1 : -1;
    });
};
