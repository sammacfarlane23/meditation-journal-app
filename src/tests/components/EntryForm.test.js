import { shallow } from 'enzyme';
import React from 'react';
import EntryForm from '../../components/EntryForm';
import entries from '../fixtures/entries';
import moment from '../__mocks__/moment';

test('should render EntryForm correctly', () => {
  const wrapper = shallow(<EntryForm />);
  expect(wrapper).toMatchSnapshot();
});

test('should render EntryForm correctly with entry data and delete button', () => {
  const wrapper = shallow(<EntryForm entry={entries[1]} canDelete={true} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid form submission', () => {
  const wrapper = shallow(<EntryForm />);
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {},
  });
  expect(wrapper.find('.form__error').length).toBeGreaterThan(0);
});

test('should set title on input change', () => {
  const value = 'This is a title';
  const wrapper = shallow(<EntryForm />);
  wrapper.find('input').at(0).simulate('change', {
    target: { value },
  });
  expect(wrapper.find('.title').props().value).toBe(value);
});

test('should set entryText on textarea change', () => {
  const value = 'This is entry text';
  const wrapper = shallow(<EntryForm />);
  wrapper.find('textarea').simulate('change', {
    target: { value },
  });
  expect(wrapper.find('textarea').props().value).toBe(value);
});

test('should call onSubmit prop for valid form submission', () => {
  const onSubmitSpy = jest.fn();
  const wrapper = shallow(
    <EntryForm entry={entries[2]} onSubmit={onSubmitSpy} />
  );
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {},
  });
  expect(wrapper.find('.form__error').length).toBe(0);
  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    title: entries[2].title,
    entryText: entries[2].entryText,
    createdAt: entries[2].createdAt,
  });
});

test('clicking edit date button should show date and time picker', () => {
  const wrapper = shallow(<EntryForm />);
  wrapper.find('.edit-button').simulate('click');
  expect(wrapper.find('SingleDatePicker').length).toBeGreaterThan(0);
  expect(wrapper.find('.input-time').length).toBeGreaterThan(0);
});

test('should set new date on date change after clicking edit date to bring up date picker', () => {
  const now = moment();
  const wrapper = shallow(<EntryForm />);
  wrapper.find('.edit-button').simulate('click');
  wrapper.find('SingleDatePicker').prop('onDateChange')(now);
  expect(wrapper.find('SingleDatePicker').props().date).toEqual(now);
});

test('should set calendar focus after clicking edit date to bring up date picker', () => {
  const focused = true;
  const wrapper = shallow(<EntryForm />);
  wrapper.find('.edit-button').simulate('click');
  wrapper.find('SingleDatePicker').prop('onFocusChange')({ focused });
  expect(wrapper.find('SingleDatePicker').props().focused).toBe(focused);
});

test('should call closeModal prop on clicking cancel button', () => {
  const closeModalSpy = jest.fn();
  const wrapper = shallow(<EntryForm closeModal={closeModalSpy} />);
  wrapper.find('.button--cancel').simulate('click');
  expect(closeModalSpy).toBeCalled();
});

test('should call onRemove prop on clicking delete button', () => {
  const onRemoveSpy = jest.fn();
  const wrapper = shallow(
    <EntryForm onRemove={onRemoveSpy} canDelete={true} />
  );
  wrapper.find('.button--delete').simulate('click');
  expect(onRemoveSpy).toBeCalled();
});

// test('should set new time on time change after clicking edit date to bring up time picker', () => {
//   const currentDate = moment().startOf('day');
//   const timeOfDay = moment.duration(2, 'hours');
//   const createdAt = moment(moment().startOf('day')).add(timeOfDay);
//   const wrapper = shallow(<EntryForm />);
//   wrapper.find('button').at(0).simulate('click');
//   wrapper.find('input').at(1).simulate('change', {
//     target: { value },
//   });
//   expect(wrapper.state('createdAt')).toEqual(value);
// });
