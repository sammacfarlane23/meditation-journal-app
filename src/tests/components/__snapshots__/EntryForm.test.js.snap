// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clicking edit date button should show date and time picker 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    className="title"
    onChange={[Function]}
    placeholder="Title (optional)"
    type="text"
    value=""
  />
  <hr
    className="horizontal-line"
  />
  <textarea
    className="text-area"
    onChange={[Function]}
    placeholder="Describe your experience here..."
    rows="10"
    value=""
  />
  <div
    className="date-edit"
  >
    <div
      className="single-date-picker"
    >
      <SingleDatePicker
        anchorDirection="left"
        customCloseIcon={null}
        customInputIcon={null}
        date={"1970-01-01T00:00:00.000Z"}
        daySize={39}
        disabled={false}
        displayFormat="DD/MM/YYYY"
        enableOutsideDays={false}
        firstDayOfWeek={null}
        focused={false}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        id="date"
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepOpenOnDateSelect={false}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPrev={null}
        numberOfMonths={1}
        onClose={[Function]}
        onDateChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableDate": [Function],
            "clearDate": "Clear Date",
            "closeDatePicker": "Close",
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month",
            "jumpToPrevMonth": "Move backward to switch to the previous month",
            "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
            select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
            "moveFocusByOneMonth": "Switch months",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
            "openThisPanel": "Open this panel",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field",
            "selectFocusedDate": "Select the date in focus",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        placeholder="Date"
        readOnly={false}
        renderCalendarInfo={null}
        renderDay={null}
        renderMonth={null}
        reopenPickerOnClearDate={false}
        required={false}
        screenReaderInputMessage=""
        showClearDate={false}
        showDefaultInputIcon={false}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={true}
      />
    </div>
    <input
      className="input-time"
      max="01:00"
      onChange={[Function]}
      type="time"
      value="01:00"
    />
  </div>
  <div
    className="button-container"
  >
    <div
      className="remove-button-container"
    />
    <div
      className="other-button-container"
    >
      <button
        className="button button--right"
      >
        Save
      </button>
      <button
        className="button button--right button--cancel"
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
    </div>
  </div>
</form>
`;

exports[`should render EntryForm correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    className="title"
    onChange={[Function]}
    placeholder="Title (optional)"
    type="text"
    value=""
  />
  <hr
    className="horizontal-line"
  />
  <textarea
    className="text-area"
    onChange={[Function]}
    placeholder="Describe your experience here..."
    rows="10"
    value=""
  />
  <div
    className="createdAt-section"
  >
    <p
      className="createdAt-section__text"
    >
      01 Jan 1970 01:00
    </p>
    <button
      className="edit-button"
      onClick={[Function]}
      type="button"
    >
      <Component />
    </button>
  </div>
  <div
    className="button-container"
  >
    <div
      className="remove-button-container"
    />
    <div
      className="other-button-container"
    >
      <button
        className="button button--right"
      >
        Save
      </button>
      <button
        className="button button--right button--cancel"
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
    </div>
  </div>
</form>
`;

exports[`should render EntryForm correctly with entry data and delete button 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    className="title"
    onChange={[Function]}
    placeholder="Title (optional)"
    type="text"
    value=""
  />
  <hr
    className="horizontal-line"
  />
  <textarea
    className="text-area"
    onChange={[Function]}
    placeholder="Describe your experience here..."
    rows="10"
    value="This is third entry"
  />
  <div
    className="createdAt-section"
  >
    <p
      className="createdAt-section__text"
    >
      13 Sep 2020 13:26
    </p>
    <button
      className="edit-button"
      onClick={[Function]}
      type="button"
    >
      <Component />
    </button>
  </div>
  <div
    className="button-container"
  >
    <div
      className="remove-button-container"
    >
      <button
        className="button button--delete"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </div>
    <div
      className="other-button-container"
    >
      <button
        className="button button--right"
      >
        Save
      </button>
      <button
        className="button button--right button--cancel"
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
    </div>
  </div>
</form>
`;

exports[`should render error for invalid form submission 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    className="title"
    onChange={[Function]}
    placeholder="Title (optional)"
    type="text"
    value=""
  />
  <hr
    className="horizontal-line"
  />
  <p
    className="form__error"
  >
    Please add an entry
  </p>
  <textarea
    className="text-area"
    onChange={[Function]}
    placeholder="Describe your experience here..."
    rows="10"
    value=""
  />
  <div
    className="createdAt-section"
  >
    <p
      className="createdAt-section__text"
    >
      01 Jan 1970 01:00
    </p>
    <button
      className="edit-button"
      onClick={[Function]}
      type="button"
    >
      <Component />
    </button>
  </div>
  <div
    className="button-container"
  >
    <div
      className="remove-button-container"
    />
    <div
      className="other-button-container"
    >
      <button
        className="button button--right"
      >
        Save
      </button>
      <button
        className="button button--right button--cancel"
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
    </div>
  </div>
</form>
`;
